### ISSUE TITLE
GPT-4 does not work with my code

### ISSUE THREAD

**Issue by irtiq7** [OpenBMB/ChatDev#253]
Hi,

Everytime I try to built a program using --model "GPT_4", I end up getting an error message

`openai.error.RateLimitError: Rate limit reached for gpt-4 in organization org-nzQ372y9RLtqI3k1M1ACHSvA on tokens per min. Limit: 10000 / min. Please try again in 6ms. Visit https://platform.openai.com/account/rate-limits to learn more.`

Could someone share what this error is and how to fix it? I came across [This](https://github.com/OpenBMB/ChatDev/pull/140) but this patch crashes. Also, I see that my money is burned in the process.

**Comment by howtobearealprogrammer** [comment]
I had encountered this too.
To fix this (it relates to our usage tiers or some such) I did the following:
To  \Chatdev-Main\camel\model_backend,py I made these changes:

In the imports section that looks like:
```
from camel.typing import ModelType
from chatdev.statistics import prompt_cost
from chatdev.utils import log_and_print_online

```
I added 
```
from tenacity import retry, wait_exponential
from time import sleep
```

Then inside the OpenAIModel class (line 52 or 53ish) I added:
`@retry(wait=wait_exponential(multiplier=30, max=100)) `
right before the def run(self ... line so that it now looks like this:

```
    @retry(wait=wait_exponential(multiplier=30, max=100)) 
    def run(self, *args, **kwargs) -> Dict[str, Any]:
```
    
    
 then inside the def run code right after the line that says 
 
`self.model_config_dict['max_tokens'] = num_max_completion_tokens`
 
and before the line starting response = openai.ChatCompletion.create etc add this:

`sleep(3)`

In my code it looks like so:
```
        self.model_config_dict['max_tokens'] = num_max_completion_tokens
        # Add delay - for when using GPT4+
        sleep(3)
        response = openai.ChatCompletion.create(*args, **kwargs,
```

- then save the file and re run your prompt.
- I disable the sleep line with a # comment if I am using gpt_3_5_turbo

**Comment by symspace** [comment]
@howtobearealprogrammer Thank you for a helpful work around. I'm also getting the error: "The filename, directory name, or volume label syntax is incorrect." When leveraging my Tier 1 GPT 4 membership with $5 uploaded credits. 
I attempted to make updates to model_backend.py per your suggestions but I keep getting the following error: 

`Traceback (most recent call last):
  File "C:\Users\zayden\ChatDev\camel\model_backend.py", line 15, in <module>
    from typing import Any, Dict
ImportError: cannot import name 'Any' from 'typing' (C:\Users\zayden\ChatDev\camel\typing.py)`

I saw it might be my regarding my Python version but I have the most up to date installed and I also successfully installed pip-23.3.1 for typing-extensions package. 
Would appreciate any advice because I'd love to successfully use chat_dev and I've spend hours trying to get it to work (.___.)'

**Comment by howtobearealprogrammer** [comment]
Symspace,
To be transparent; ChatDev is the first time I have dealt with Python. My work world is a totally different beast I can't talk about because NDAs and my Personal world involves arduino and robots.

(I don't think it is a `pip` problem - but breaking changes in the `python` versions and if you use something to allow you to run 3.9._latest_ ChatDev should run. It did *not* run for me with the _latest_ version of Python)

When I first installed ChatDev this is the chain of events on a Windows machine:

- Installed absolute latest version of python (not 3.9.x - this is where the fun actually begun for me and the clue lies)
- Added all the "recommended" python extensions to vscode
- downloaded (because forking scares me) the ChatDev repo
- followed the quick start instructions in the wiki precisely (omg the authors are fantastic humans?)
- got Anaconda installed
- got conda to work from Powershell instance (path variable etc)
- "ran" the wiki's conda lines (if that makes sense)
- failed when I went to do the first python run.py etc.
- swore. went back and read the details - PYTHON 3.9.... 

got something called `pyenv` (? I have had a long day today) - but the version for Windows. then also downloaded the Official Python 3.9.13 install etc, following the ?`pyenv`? guide

Used ?`pyenv`? to set the GLOBAL version of python to 3.9.13 (there is stuff in the install instructions on how to do this.. I did read it reasonably slowly with a glass of pinot noir)

Then I re launched the PS Shell, did the "Python -V" and got back 3.9.13. 
- THIS IS IMPORTANT (at least for me it was) - If Powershell does not think it is running Python 3.9.x then I could not get Chat Dev to run.

TL:DR - Check you're using Python 3.9.something - this avoids breaking changes (especially if you don't like hunting for that kinda mess)

Too long, really didn't read - yes we puts that nearly at the end 🙂

The next thing that broke for me was the `$env:OPENAI_API_KEY:` _*YOUR SECRET API KEY GOES HERE*_ bit - It is mentioned in the instructions, I missed it first time, then I misspelt OPENAI but you *DO.* *NOT.* want to store this as a static windows environment variable - it gives access to your dollars attached to an astounding LLM - I take the effort to put this in manually for each terminal/PowerShell session I am using.
